app/database/build_store.py

```diff
@@ -0,0 +1,307 @@
+        build._artifacts_zip_file = q_build.artifacts_zip_file
+        build._error_message = q_build.error_message
+        build._postbuild_tasks_are_finished = bool(int(q_build.postbuild_tasks_are_finished))
+        build.setup_failures = q_build.setup_failures
+        build._timing_file_path = q_build.timing_file_path
+
+        # Manually set the state machine timestamps
+        build._state_machine._transition_timestamps = {
+            BuildState.QUEUED: q_build.queued_ts,
+            BuildState.FINISHED: q_build.finished_ts,
+            BuildState.PREPARED: q_build.prepared_ts,
+            BuildState.PREPARING: q_build.preparing_ts,
+            BuildState.ERROR: q_build.error_ts,
+            BuildState.CANCELED: q_build.canceled_ts,
+            BuildState.BUILDING: q_build.building_ts
+        }
+        build._state_machine._fsm.current = BuildState[q_build.state]
+
+        build_artifact = BuildArtifact(q_build.build_artifact_dir)
+
+        directories = []
+        for directory in q_failed_artifact_directories:
+            directories.append(directory.failed_artifact_directory)
+        build_artifact._failed_artifact_directories = directories
+
+        pairs = []
+        for pair in q_failed_subjob_atom_pairs:
+            pairs.append((pair.subjob_id, pair.atom_id))
+        build_artifact._q_failed_subjob_atom_pairs = pairs
+
+        build._build_artifact = build_artifact
+
+        atoms_by_subjob_id = {}
+        for atom in q_build_atoms:
+            atoms_by_subjob_id.setdefault(atom.subjob_id, [])
+            atoms_by_subjob_id[atom.subjob_id].append(Atom(
+                atom.command_string,
+                atom.expected_time,
+                atom.actual_time,
+                atom.exit_code,
+                atom.state,
+                atom.atom_id,
+                atom.subjob_id
+            ))
+
+        subjobs = OrderedDict()
+        for subjob in q_build_subjobs:
+            atoms = atoms_by_subjob_id[subjob.subjob_id]
+            # Add atoms after subjob is created so we don't alter their state on initialization
+            subjob_to_add = Subjob(build_id, subjob.subjob_id, build.project_type, job_config, [])
+            subjob_to_add._atoms = atoms
+            subjob_to_add.completed = subjob.completed
+            subjobs[subjob.subjob_id] = subjob_to_add
+        build._all_subjobs_by_id = subjobs
+
+        # Place subjobs into correct queues within the build
+        build._unstarted_subjobs = Queue(maxsize=len(subjobs))
+        build._finished_subjobs = Queue(maxsize=len(subjobs))
+        for _, subjob in subjobs.items():
+            if subjob.completed:
+                build._finished_subjobs.put(subjob)
+            else:
+                build._unstarted_subjobs.put(subjob)
+
+        # Spend preparation coin if this build is already completed
+        if q_build.completed:
+            build._preparation_coin.spend()
```

Comment:
Overall this reconstruction seems complex and is messing around with the internals of a Build object that nothing besides the Build class itself should really be messing with. There seems to be a lot here around recreating the state of in-progress builds, which is added complexity that I think we should avoid on this first pass (if not forever).

Refactoring the Build class is already very hard because of its complexity (and so it _should_ be refactored to make it simpler). I don't want to make it harder to refactor, but I think the amount of internals touched by this persistence logic will make it much harder.

Really, the main requirement is to allow users to get the state of previous builds after the master is restarted. To me, that means the only data the database needs to contain is what is exposed by the API. We don't need to recreate a Build object exactly as it existed before the restart -- we just need to make sure the data that users are interested in is available.